Write blog post about faster `rand` generators (#1)
RaimoNiskanen/erlang-org: _posts/2022-04-20-faster-rand.md

---
layout: post
title:  Fast random integers
tags:   BEAM JIT
author: Raimo Niskanen
---

When you need "random" integers, and it is essential
to generate them fast and cheap; then maybe the full featured
Pseudo Random Number Generators in the `rand` module are overkill.
This blog post will dive in to new additions to the
said module, how the Just-In-Time compiler optimizes them,

> known tricks, and and tries to compare these apples and potatoes.

Reviewer: ```suggestion
known tricks, and tries to compare these apples and potatoes.
```

Speed over quality?
-------------------

> The Pseudo Random Number Generators implemented in

Reviewer: ```suggestion
The Pseudo Random Number Generators (PRNGs) implemented in
```

the `rand` module offers many useful features such as
repeatable sequences, non-biased range generation,

> any size (bignum) range, non-overlapping sequences,

Reviewer: Here you mention "bignum" without defining it. I suggest that you delete this mention and defines "bignum" a little bit later.
 
```suggestion
any size range, non-overlapping sequences,
```

generating floats, normal distribution floats, etc.
Many of those features are implemented through

> a plug-in framework, whith a performance cost.

Reviewer: ```suggestion
a plug-in framework, with a performance cost.
```

The different algorithms offered by the `rand` module are selected
to have excellent statistical quality and to perform well

> in serious PRNG tests, (see section [PRNG tests]).

Reviewer: ```suggestion
in serious PRNG tests (see section [PRNG tests]).
```

> Most of these algorithms are designed for machines with

Reviewer: Somewhere here could be a good place to define what bignums are and also explain why `rand` uses 58-bit integers (why not 59-bit integers).

64-bit arithmetic (unsigned), but in Erlang such integers
become bignums and almost an order of magnitude slower
to handle than immediate integers.

The algorithms that perform well in Erlang are the ones

> that have been redesigned to work on 58-bit words.

Reviewer: ```suggestion
that have been redesigned to work on 58-bit integers.
```

But still, when executed in Erlang, they are far from
as fast as their C origins.  Achieving good PRNG quality
costs much more in Erlang than in C.  In the section
[Measurement results] we see that the algorithm `exsp`
that boasts sub-ns speed in C needs 17 ns in Erlang.

32-bit Erlang is a sad story in this regard.  The bignum limit

> on such an Erlang system is so low that designing a PRNG

Reviewer: How low?

Author: You tell me ;-)  `erlang:phash2/1` returns 27 bits.  Is 2^27 - 1 the upper limit for a positive immediate?

Reviewer: Yes.

Author: Great!  Then we would have to use 26-bit integers.  An LCG would have to have a multiplication constant of 13 bits.  So size and period about 8192.  Ridiculous!

that uses non-bignums has to use such a small word size
that the generator becomes too bad to be used.
The known trick `erlang:phash2(erlang:unique_integer(), Range)`
is still fairly fast, but all `rand` generators work exactly the same
as on a 64-bit system, hence operates on bignums so they are much slower.

If your application needs a "random" integer for an non-critical
purpose such as selecting a worker, choosing a route, etc,
and performance is much more important than repeatability
and statistical quality, what are then the options?

Suggested solutions
-------------------

* [Use `rand` anyway]
* [Write a BIF]
* [Write a NIF]
* [Use the system time]

> * [Hash an "unique" value]

Reviewer: ```suggestion
* [Hash a "unique" value]
```
(The "u" is **pronounced** with a consonant sound; thus "a" instead of "an".)

Author: Although I know the rule, when I get tired i sometimes mistake a fricative for a vowel...

* [Write a simple PRNG]

Reasoning and measurement results are in the following sections,
but, in short:

* Writing a NIF, we deemed, does not achieve a performance worth the effort.
* Neither does writing a BIF.  But, ... a BIF (and a NIF, maybe) could
  implement a combination of performance and quality that cannot
  be achieved in any other way.  If a high demand on this combination
  would emerge, we could reconsider this decision.
* Using the system time is a bad idea.
* `erlang:phash2(erlang:unique_integer(), Range)` has its use cases.
* We have implemented two simple PRNGs to fill the niche of non-critical
  but very fast number generation: `lcg35` and `mcg35`.

### Use `rand` anyway

Is `rand` slow, really?  Well, perhaps not considering what it does.

In the [Measurement results] at the end of this text,
it shows that generating a good quality random number using
the `rand` module's default algorithm is done in 43 ns.

Generating a number as fast as possible (`rand:lcg35/1`) can be done
in just above 3 ns, but that algorithm has got pretty bad

> statistical quality by to today's standards.  See section [PRNG tests].

Reviewer: ```suggestion
statistical quality by today's standards.  See section [PRNG tests].
```

Using a good quality algorithm instead (`rand:exsp_next/1`) takes 16 ns,
if you can store the generator's state in a loop variable.

If you can not store the generator state in a loop variable
there will be more overhead, see section [Storing the state].

Now, if you also need a number on an awkward range, as in not much smaller
than the generator's size, you might have to implement a reject-and-resample
loop, or must concatenate numbers.

The overhead of code that has to implement this much of the features
that the `rand` module already offers will easily approach

> it's 26 ns overhead, so often there is no point in

Reviewer: ```suggestion
its 26 ns overhead, so often there is no point in
```

re-implementing this wheel...

### Write a BIF

There was been a discussion thread on Erlang Forums:
[Looking for a faster RNG].  Triggered by this the user "potatosallad"
wrote an experimental BIF.

The suggested BIF `erlang:random_integer(Range)` offered
no repeatability, generator state per scheduler, guaranteed

> sequence separatiton between schedulers, and high generator

Reviewer: ```suggestion
sequence separation between schedulers, and high generator
```

quality.  All this due to using one of the good generators from
the `rand` module, but now written in C in the BIF.

The performance was comparable to the `mcg35` generator, so a bit slower
than the (fastest) 'lcg53' generator, but with top of the line quality.
See section [Measurement results].

Questions arised regarding maintenance burden, what more to implement, etc.
For example we probably also would need `erlang:random_integer/0`,
`erlang:random_float/0`, and some system info
to get the generator bit size...

A BIF could achieve good performance on a 32-bit system, if it there
would return a 27-bit integer, which became another open question.
Should a BIF generator be platform independent with respect to
generated numbers or with respect to performance?

### Write a NIF

"potatosallad" also wrote a NIF, since we (The Erlang/OTP team)
suggested that it could have good enough performance.

Measurements, however, showed that the overhead is significantly larger
than for a BIF.  Although the NIF used the same trick as the BIF to store
the state in thread specific data it ended up with the same
performance as `erlang:phash2(erlang:unique_integer(), Range)`,

> which is about 2..3 times slower than the BIF.

Reviewer: `..` is not an operator in Erlang or C, so it might not be familiar to all readers.

```suggestion
which is about 2 to 3 times slower than the BIF.
```

As a speed improvement we tried was to have the NIF generate
a list of numbers, and use that list as a cache in Erlang.
The performance with such a cache was as fast as the BIF,
but introduced problems such as that you would have to decide
on a cache size, the application would have to keep the cache on the heap,
and when generating on a range the application would have to know
in advance the range for the whole cache.

A NIF could like a BIF also achieve good performance on a 32-bit system,

> with the same open question - platform independent numbers or performance?

Reviewer: ```suggestion
with the same open question &mdash; platform independent numbers or performance?
```

Author: What is an mdash?

Reviewer: It is a horizontal line &mdash; roughly as long as an "m" used to mark a break in a sentence. See [Dash](https://en.wikipedia.org/wiki/Dash) and also compare the Swedish [Tankstreck](https://sv.wikipedia.org/wiki/Tankstreck), which is shorter but used in the same way as an  em dash in English.

### Use the system time

One suggested trick is to use `os:system_time(microseconds)` to get
a number.  The trick has some pecularities:
* When called repeatedly you might get the same number several times.
* The resolution is system dependent, so on some systems you get
  the same number even more several times.
* Time can jump backwards and repeat in some cases.
* Historically it has been a bottleneck, especially on virtualized
  platforms.  Getting the OS time is harder then expected.

See sectom [Measurement results] for the performance for this "solution".

> ### Hash an "unique" value

Reviewer: ```suggestion
### Hash a "unique" value
```

The best combination would most certainly be
`erlang:phash2(erlang:unique_integer(), Range)` or
`erlang:phash2(erlang:unique_integer())` which is slightly faster.

`erlang:unique_integer/0` is designed to return a unique integer
with a very small overhead.  It is hard to find a better candidate
for an integer to hash.

`erlang:phash2/1,2` is the current generic hash function for Erlang terms.
It has a default return size well suited for 32-bit Erlang systems,
and it has a `Range` argument.  The range capping is done with a simple
`rem` in C (`%`) which is much faster than in Erlang.  This works good
 only for ranges much smaller than 32-bit.

See section [Measurement results] for the performance for this solution.

### Write a simple PRNG

To be fast, the implementation of a PRNG algorithm cannot

> execute many operations operations.  The operations have to be

Reviewer: ```suggestion
execute many operations.  The operations have to be
```

on immediate values (not bignums), and the state as well
as the returned number also have to be immediate values.
This seriously limits how powerful algorithms that can be used.

A Linear Congruential Generator is one of the most researched
and oldest PRNG:s.  It is implemented like this (in Erlang):
``` erlang
    X1 = (A * X0 + C) rem P
```
For this to produce a sequence of numbers that appear random,

> there are a number of requirements on the constants A, C and P.

Reviewer: ```suggestion
there are a number of requirements on the constants `A`, `C` and `P`.
```

One criteria for statistical quality is the spectral score,
see section [Spectral score].

To get a decent spectral score `A` cannot be too small, not (much)
smaller than the square root of `P`.  And `(A * X0 + C)` produces a number
as large as `A * P`, which must not become a bignum.

The `rem` operation is expensive, but if `P` is a power of 2,
we can use `band (P-1)` instead, which is much faster.

To get a full sequence, that is, to use all numbers in the state range,
there are more restrictions imposed on `A`, `P` and `C`, but we will
not dig deeper into this field and instead use established research.

There is a classical (1999) paper by Pierre L'Ecuyer:

> "Tables of linear congruential generators of different sizes

Reviewer: Is there an online version that you can link to?

and good lattice structure".  Lattice structure is the property
that the spectral score measures.

From that paper I selected two generators that avoids bignums:
``` erlang
    X1 = (15319397 * X0 + 15366142135) band ((1 bsl 35)-1)
```
which I named `lcg35`, and
``` erlang
    X1 = (185852 * X0) rem ((1 bsl 35)-31)
```
which I named `mcg35`

`lcg35` is a power of 2 generator with an odd addition constant

> and a state interval of `0 =< X < 2^35`.

Reviewer: `^` is not defined in Erlang and it means XOR in C.

I suggest using proper mathematical notation like this: 0 &leq; X < 2 <sup>35</sup>.

```suggestion
and a state interval of 0 &leq; X < 2 <sup>35</sup>.
```

Author: The Markdown source becomes unreadable which goes against the idea of Markdown...

Reviewer: Yes, but this is a blog post. No one will read the Markdown source.

`mcg35` is a prime modulus multiplicative generator (Lehmer generator)

> with a state interval of `1 =< X < 2^35-31`.  Since the modulus,

Reviewer: ```suggestion
with a state interval of 1  &leq; X < 2 <sup>35</sup>-1.  Since the modulus,
```

operator, but it is noticably slower than a bitwise operation.

Other tricks are possible, for example if you need numbers
on the range 0..999 you may use bitwise operations to get
a number 0..1023, and if too high re-try, which actually
may be faster on average than using `rem`.

### Spectral score

The specral score of a generator, measures how a much a sequence of numbers
from the generator are unrelated.  A sequence of N numbers are interpreted
as an N-dimensional vector and the spectral score for dimension N is a measure
on how evenly these vectors are distributed in an N-dimensional (hyper)cube.

`os:system_time(microseconds)` simply increments so it should have
a lousy spectral score.

`erlang:phash2(erlang:unique_integer(), Range)` has got unknown
spectral score, since that is not part of the math behind a hash function.
But a hash function is designed to distribute the hash value well
for any input, so we can assume that the statistical
distribution of the numbers is decent and "random".

All PRNG generators discussed here have got good spectral scores.

### PRNG tests

There are test frameworks that tests the statistical properties

> of PRNG:s, such as the TestU01 framework, or PracRand.

Reviewer: Can you link to them?

Seeding is related to predictability.  If you can guess
the seed you know the generator output.

The seed is generator dependent and how to create a good
seed usually takes much longer than generating a number.
Sometimes the seed and its predictability is so unimportant
that a constant may be used.   If a generator instance
generates just a few numbers per seeding, then seeding
can be the harder problem.

`erlang:phash2(erlang:unique_integer(), Range)` is pre-seeded,
or rather cannot be seeded, so it has no seeding cost, but can
on the other hand be rather predictable, if it is possible to estimate
how many unique integers that have been generated since node start.

The default seeding in the `rand` module uses a combination
of a hash value of the node name, the system time,
and `erlang:unique_integer()`, to create a seed,
which is hopefully sufficiently unpredictable.

The suggested NIF and BIF solutions would also need
a way to create a good enough seed, where "good enough"
is hard to put a number on.

JIT optimizations
-----------------

The speed of the newly implemented `lcg53` and `mcg53` algorithms
is much thanks to the recent [type-based optimizations] in the compiler

> and the Just-In-Time compliling .BEAM code loader.

Reviewer: ```suggestion
and the Just-In-Time compiling BEAM code loader.
```

### Without type-based optimization

This is the Erlang code for the `lcg35` generator:
``` erlang
lcg35(X0) ->
    (15319397 * X0 + 15366142135) band ((1 bsl 35)-1).
```
which compiles to (Erlang .BEAM assembler, erlc -S rand.erl):
``` erlang
    {gc_bif,'*',{f,0},1,[{x,0},{integer,15319397}],{x,0}}.
    {gc_bif,'+',{f,0},1,[{tr,{x,0},number},{integer,15366142135}],{x,0}}.
    {gc_bif,'band',{f,0},1,[{tr,{x,0},number},{integer,34359738367}],{x,0}}.
```

> when loaded by the JIT (x86) (erl +JDdump true) the machine code becomes:

Reviewer: ```suggestion
when loaded by the JIT (x86) (`erl +JDdump true`) the machine code becomes:
```

```

> # i_test_yield

Reviewer: I suggest omitting the `i_test_yield` instruction and only show the instructions relevant to the example.

Author: Ok.  My mistake - I thought that was part of the multiplication...

    lea rdx, qword ptr [lcg35/1+24]
    dec r14
    jle L423
# i_times_jssd
    mov rsi, qword ptr [rbx]
    mov edx, 245110367
# is the operand small?
    mov edi, esi
    and edi, 15
    cmp edi, 15
    short jne L1891
```

> Above was a test for if X0 is a small integer.

Reviewer: ```suggestion
Above was a test for `{x,0}` being a small integer.
```

Below is a multiplication that has to check for overflow
```
# mul with overflow check, imm RHS
    mov rax, rsi
    mov rcx, 15319397
    and rax, -16
    imul rax, rcx
    short jo L1891
    or rax, 15
    short jmp L1890
```

> If X0 was not a small integer the following fallback code is used

Reviewer: I suggest using `{x,0}` (the same as in the BEAM code shown earlier) instead of `X0`.

```suggestion
If `{x,0}` was not a small integer the following fallback code is used
```

that handles multiplication of any term.
```
L1891:
    call 140408504026912
L1890:
    mov qword ptr [rbx], rax
```
The following addition knows that the argument is a number,
so it can use a simplified test for if the argument
is a small integer and then it has a fallback for other terms
which could be a bignum or a float.
```
# i_increment_SWd
    mov rsi, qword ptr [rbx]
    mov rdx, 245858274160
# simplified test for small operand since it is a number
    mov rax, rsi
    test al, 1
    short je L1892
    add rax, rdx
    short jno L1893
L1892:
    call 140408504025560
L1893:
    mov qword ptr [rbx], rax
```
Here is a mask operation to 35 bits that also
knows that the argument is a number.
```
# i_band_ssjd
    mov rsi, qword ptr [rbx]
    mov rax, 549755813887
# simplified test for small operand since it is a number
    test esi, 1
    short je L1894
    and rax, rsi
    short jmp L1895
L1894:
    call 140408504023680
L1895:
    mov qword ptr [rbx], rax
# return
    dec r14
    jl L424
    ret
```

### With type-based optimization

When the compiler can figure out type information about the arguments

> it can emit more effective code.  One would like to add a guard

Reviewer: ```suggestion
it can emit more efficient code.  One would like to add a guard
```

that restricts the argument to a 58 bit integer, but unfortunately
the compiler cannot yet make use of such a guard test.

But adding a redundant input bit mask to the Erlang code puts the compiler
on the right track.  This is a kludge, and will only be used
until the compiler has been improved to deduce the same information

> from a guard instead:

Reviewer: ```suggestion
from a guard instead.
```

Erlang code; note the first redundant mask to 35 bits:
``` erlang
lcg35(X0) ->
    X = X0 band ((1 bsl 35)-1),
    (15319397 * X + 15366142135) band ((1 bsl 35)-1).
```
The .BEAM assembler now becomes:
``` erlang
    {gc_bif,'band',{f,0},1,[{x,0},{integer,34359738367}],{x,0}}.
    {gc_bif,'*',
            {f,0},
            1,
            [{tr,{x,0},{t_integer,{0,34359738367}}},{integer,15319397}],
            {x,0}}.
    {gc_bif,'+',
            {f,0},
            1,
            [{tr,{x,0},{t_integer,{0,526370472860204699}}},
             {integer,15366142135}],
            {x,0}}.
    {gc_bif,'band',
            {f,0},
            1,
            [{tr,{x,0},{t_integer,{15366142135,526370488226346834}}},
             {integer,34359738367}],
            {x,0}}.
```
Note the `{t_integer,Range}` type information that instead of just declaring

> that the argument is a number, now states that it is an integer on

Reviewer: ```suggestion
that the argument is a number, now states that it is an integer in
```

The first two are the warm-up and overhead measurements.
The measured overhead is subtracted from all measurements
after the "overhead:" line.  Note that the measured overhead
is 3.3 ns which matches pretty well that `exsss` when adjusted
for overhead got 3.6 ns shorter time than on the warm-up run.

`mcg35_inline`, `lcg35_inline`, `exsp_inline` and `system_time`
all use `(X rem 10000) + 1` to achieve the desired range.
This operation is expensive, which we will see when comparing
with the next section.

`erlang:phash2/2` has got a range argument, that performs
the `rem 10000` operation in the BIF, which is cheap,
as we also will see when comparing with the next section.

```
RNG uniform integer 32 bit performance
                   exsss:     54.4 ns    100.0%
                    exsp:     49.4 ns     90.8%
            lcg35_inline:      3.0 ns      5.6%
             exsp_inline:     16.8 ns     30.9%
           unique_phash2:     22.0 ns     40.4%
             system_time:     23.6 ns     43.4%
```
In this section `lcg35_inline`, `exsp_inline` and `system_time`
use bit operations such as `X band 16#ffffffff` or `X bsr 3`
to achieve the desired range, and now we see that this is

> about 5..10 ns faster than for the `rem` operation

Reviewer: ```suggestion
about 5 to 10 ns faster than for the `rem` operation
```

`erlang:unique_integer/0` is designed to return an unique integer
with a very small overhead.  It is hard to find a better candidate
for an integer to hash.

`erlang:phash2/1,2` is the current generic hash function for Erlang terms.
It has a default return size well suited for 32-bit Erlang systems,
and it has a `Range` argument.  The range capping is done with a simple
`rem` in C (`%`) which is much faster than in Erlang.  This works good
 only for ranges much smaller than 32-bit.

Alas this solution does not perform well in [PRNG tests].

See section [Measurement results] for the performance for this solution.

### Write a simple PRNG

To be fast, the implementation of a PRNG algorithm cannot
execute many operations.  The operations have to be
on immediate values (not bignums), and the state as well
as the returned number also have to be immediate values.
This seriously limits how powerful algorithms that can be used.

A Multiply With Carry generator is one of the classical PRNG:s
and is a special form of a Muliplicative Contruential Generator.
It is a well researched PRNG, and can be implemented
like this (in Erlang):
``` erlang

>     C   = CX0 bsl B,

Reviewer: ```suggestion
    C   = CX0 bsr B,
```

you generate and on how large range you generate, or it may affect
the simulation for example that specific numbers do not reoccur.
If you have pulled 3 aces from a deck you know there is only one left.

Some applications may be sensitive to the generator period,
while others are not, and this needs to be considered.

### Size

The size of the new fast generators is about 35 bits, which is both
the generator's state size (period) and value size.
The good quality generators in the `rand` module has got a size
(value size) of 58 bits.

If you need numbers in a power of 2 range then you can
simply mask out or shift down the required number of bits,
depending on if the generator is known to have weak high or low bits.

If the range you need is not a power of 2, but still
much smaller than the generator's size you can use the `rem`
operator, but it is noticably slower than a bitwise operation.

Other tricks are possible, for example if you need numbers
in the range 0 through 999 you may use bitwise operations to get
a number 0 through 1023, and if too high re-try, which actually
may be faster on average than using `rem`.

### Spectral score

> The specral score of a generator, measures how a much a sequence of numbers

Reviewer: ```suggestion
The spectral score of a generator, measures how a much a sequence of numbers
```

Author: Also: "how a much a" -> "how much a"

which is hopefully sufficiently unpredictable.

The suggested NIF and BIF solutions would also need
a way to create a good enough seed, where "good enough"
is hard to put a number on.

JIT optimizations
-----------------

The speed of the newly implemented `mwc59` algorithm
is partly thanks to the recent [type-based optimizations] in the compiler
and the Just-In-Time compiling BEAM code loader.

### With no type-based optimization

This is the Erlang code for the `mwc59` generator:
``` erlang
mwc59(CX) ->
    16#7f17555 * (CX band ((1 bsl 32)-1)) + (CX bsr 32).
```
which compiles to (Erlang BEAM assembler, `erlc -S rand.erl`),
using the `no_type_opt` flag:
``` erlang
    {gc_bif,'bsr',{f,0},1,[{x,0},{integer,32}],{x,1}}.
    {gc_bif,'band',{f,0},2,[{x,0},{integer,4294967295}],{x,0}}.
    {gc_bif,'*',{f,0},2,[{x,0},{integer,133264725}],{x,0}}.
    {gc_bif,'+',{f,0},2,[{x,0},{x,1}],{x,0}}.
```
when loaded by the JIT (x86) (`erl +JDdump true`) the machine code becomes:
```
# i_bsr_ssjd
    mov rsi, qword ptr [rbx]
# is the operand small?
    mov edi, esi
    and edi, 15
    cmp edi, 15
    short jne L2271
```
Above was a test if `{x,0}` is a small integer and if not

> the fallback at `L1816` is called to handle any term.

Reviewer: ```suggestion
the fallback at `L2271` is called to handle any term.
```

> Here follows the machine code shift left, Erlang `bsl 32`,

Reviewer: ```suggestion
Here follows the machine code for shift right, Erlang `bar 32`,
```

> x86 `sar rax 32`, and a skip over the fallback code.

Reviewer: ```suggestion
x86 `sar rax, 32`, and a skip over the fallback code.
```

off the high 3 bits gives 8 possibilities for all values but the top 31
and zero, that only get 7 possibilities.  So the probability for
a zero or one of the top 31 values is 7/8 of the probability for all others.

Note that it is unwise to use more than say half the bits
corresponding to a generator's period anyway. See section [Quality],
about that and other differences between these generators.

As gap filler between really fast with low quality, and full featured,
an internal function in `rand` has been exported: `rand:exsp_next/1`.
This function implements Xoroshiro116+ and is used from the
`rand` plug-in framework as algorithm `exsp`.  It has been
exported so it is possible to get good quality without
the plug-in framework overhead, for applications that do not
need any framework features.

See section [Measurement results] for speed comparisions.

Quality
-------

There are many different aspects of a PRNG:s quality.
Here are some.

### Period

`erlang:phash2(erlang:unique_integer(), Range)` has, conceptually,
an infinite period, since the time it will take for it to repeat
is assumed to be longer than the Erlang node will survive.

> For the new fast generators the period it is about 2<sup>35</sup>`,
> and for the good ones in `rand` it is at least `2<sup>116</sup>&nbsp;-&nbsp;1,
> which is a huge difference and also much more than what can be consumed

Reviewer: Remove stray backticks which caused strange formatting:

```suggestion
For the new fast generators the period it is about 2<sup>35</sup>,
and for the good ones in `rand` it is at least 2<sup>116</sup>&nbsp;-&nbsp;1,
which is a huge difference and also much more than what can be consumed
```

Author: Bummer!  I missed some!
